### PR TITLE
[10_2_X] MXNet-based implementation of the ParticleNet tagger

### PR DESCRIPTION
**Pull request by hqucms**
Backport of https://github.com/cms-sw/cmssw/pull/28902. 
For analysis of the full Run2 data and NanoAOD production.

Needs:
 - DNN model: https://github.com/cms-data/RecoBTag-Combined/pull/26
 - MXNet 1.5: https://github.com/cms-sw/cmsdist/pull/5591

We should also integrate https://github.com/cms-data/RecoBTag-Combined/pull/24 together with the update to MXNet 1.5.

### CODE REVIEW

**Comment by silviodonato** [comment]
@hqucms  could you explain in the PR description why you need this backport?

**Comment by hqucms** [comment]
@silviodonato We would like to have this for analysis of the full Run2 data and NanoAOD production. I updated the PR description with that.

**Comment by silviodonato** [comment]
@hqucms could you confirm the last two commits make this PR identical to #28902 ?

**Comment by silviodonato** [comment]
please test with cms-sw/cmsdist#5591

**Comment by hqucms** [comment]
> @hqucms could you confirm the last two commits make this PR identical to #28902 ?

@silviodonato Yes -- this backport is identical as https://github.com/cms-sw/cmssw/pull/28902 except that the tagger is not enabled in the MiniAOD step to preserve the event content.

**Comment by slava77** [comment]
@cmsbuild please test

to refresh the comparisons

**Comment by slava77** [comment]
please abort test

**Comment by slava77** [comment]
please test with cms-sw/cmsdist#5591

to refresh the test outputs

**Review by slava77 — COMMENTED** [review]
I'm missing a change in PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py, which would be added via run2_miniAOD_devel.
Is this intentional?

**Comment by slava77** [comment]
@hqucms 
now that https://github.com/cms-sw/cmsdist/pull/5591 made it to the 10_2_X IB, I wanted to proceed further.

The data external, now present in the master and tagged as RecoBTag-Combined V01-03-00  should not be applied in full, we need a targeted update on top of V01-01-03 currently in the head of 10_2_X. Some more creative tag name will be needed together with a separate branch for the external data files other than the cms-data/RecoBTag-Combined:master.

**Comment by hqucms** [comment]
> @hqucms
> now that [cms-sw/cmsdist#5591](https://github.com/cms-sw/cmsdist/pull/5591) made it to the 10_2_X IB, I wanted to proceed further.
> 
> The data external, now present in the master and tagged as RecoBTag-Combined V01-03-00 should not be applied in full, we need a targeted update on top of V01-01-03 currently in the head of 10_2_X. Some more creative tag name will be needed together with a separate branch for the external data files other than the cms-data/RecoBTag-Combined:master.

Maybe @smuzaffar @mrodozov could help with creating the tag for `cms-data/RecoBTag-Combined`? What we need is to apply https://github.com/cms-data/RecoBTag-Combined/pull/26 on top of V01-01-03 (currently in the head of 10_2_X). Thanks!

**Comment by hqucms** [comment]
> I'm missing a change in PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py, which would be added via run2_miniAOD_devel.
> Is this intentional?

@slava77 Is there any re-MiniAOD foreseen w/ 10_2_X? If so I can add it via `run2_miniAOD_devel`.

File: PhysicsTools/PatAlgos/python/tools/jetTools.py [hunk 1]
```diff
@@ -644,30 +644,16 @@ def setupBTagging(process, jetSource, pfCandidates, explicitJTA, pvSource, svSou
                 if pfCandidates.value() == 'packedPFCandidates':
                     # case 1: running over jets whose daughters are PackedCandidates (only via updateJetCollection for now)
                     jetSrcName = jetSource.value().lower()
-                    if 'updated' in jetSrcName:
-                        puppi_value_map = ""
-                        vertex_associator = ""
```
[the remark this answers is not in the document]
Comment (hqucms) [reply PhysicsTools/PatAlgos/python/tools/jetTools.py:649]: These updates are needed as we use the same `TagInfo` producer for `DeepBoostedJet` and `ParticleNet`, and `ParticleNet` requires them. 
(Those changes were originally introduced to master in https://github.com/cms-sw/cmssw/pull/28035 and were not back ported to 10_2_X.)

**Comment by mrodozov** [comment]
@hqucms I'm thinking :) So you need 
to apply cms-data/RecoBTag-Combined#26 on top of V01-01-03 
and use it for 10_2 , excluding changes from releases 
V01-01-04
V01-01-05
V01-02-00
V01-02-01
and including only V01-01-03 ?

**Comment by hqucms** [comment]
> @hqucms I'm thinking :) So you need
> to apply [cms-data/RecoBTag-Combined#26](https://github.com/cms-data/RecoBTag-Combined/pull/26) on top of V01-01-03
> and use it for 10_2 , excluding changes from releases
> V01-01-04
> V01-01-05
> V01-02-00
> V01-02-01
> and including only V01-01-03 ?

@mrodozov Yes. Actually we should also include https://github.com/cms-data/RecoBTag-Combined/pull/24 to match the update of MXNet to 1.5.0.

**Comment by slava77** [comment]
> > I'm missing a change in PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py, which would be added via run2_miniAOD_devel.
> > Is this intentional?
> 
> @slava77 Is there any re-MiniAOD foreseen w/ 10_2_X? If so I can add it via `run2_miniAOD_devel`.

the _devel modifier is meant to be a catch-all for new analysis recipes.
The actual coordinated campaign will have a more specifically named modifier.
Since NanoAOD is mentioned in this PR description,
perhaps @peruzzim can comment if there is some nano-related modifier already suitable for an upcoming campaign if this PR taggers were to be included.

**Comment by mrodozov** [comment]
Yo can test with  pr28 after which we have to merge also 29 to restore the state from before 28 (this is more like not to self).
Even better, we don't merge the two prs but manually push changes and create tags to trick the auto system

**Comment by slava77** [comment]
> @mrodozov Yes. Actually we should also include [cms-data/RecoBTag-Combined#24](https://github.com/cms-data/RecoBTag-Combined/pull/24) to match the update of MXNet to 1.5.0.

@hqucms 
is cms-data/RecoBTag-Combined#24 required to correct physics or fix a problem at run time?
I was running wf 136.8311 with BoostedJetTags included using run2_miniAOD_devel and they run OK in CMSSW_10_2_X_2020-03-30-1100 baseline (with a fix from #29376).

**Comment by slava77** [comment]
> > @mrodozov Yes. Actually we should also include [cms-data/RecoBTag-Combined#24](https://github.com/cms-data/RecoBTag-Combined/pull/24) to match the update of MXNet to 1.5.0.
> 
> @hqucms
> is [cms-data/RecoBTag-Combined#24](https://github.com/cms-data/RecoBTag-Combined/pull/24) required to correct physics or fix a problem at run time?
> I was running wf 136.8311 with BoostedJetTags included using run2_miniAOD_devel and they run OK in CMSSW_10_2_X_2020-03-30-1100 baseline (with a fix from #29376).

responding to self: there are warnings of kind
``` Loading symbol saved by previous version v1.2.1. Attempting to upgrade...```
OK then

**Comment by slava77** [comment]
>  perhaps @peruzzim can comment if there is some nano-related modifier already suitable for an upcoming campaign if this PR taggers were to be included.

based on details in https://gitlab.cern.ch/cms-nanoAOD/nanoaod-doc/-/wikis/Releases/NanoAODv7
the latest nanoAOD production is using existing old modifiers (defined more than a year ago).
This is in contrast to reco/miniAOD production where existing modifiers in the old production releases are to preserve the behavior. 
Thus, I think that here ```run2_miniAOD_devel``` can enable the particleNet processing.

**Comment by slava77** [comment]
test parameters
- pull_request = cms-data/RecoBTag-Combined#29

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> ERROR: There was an issue generating parameters for
> cmsBuild '--source' flag for spec file data-RecoBTag-Combined from cms-data/RecoBTag-Combined repo.

@smuzaffar @mrodozov 
it looks like something is not right.
IIUC, the console for this build is https://cmssdt.cern.ch/jenkins/job/ib-run-pr-tests/5525/console

**Comment by smuzaffar** [comment]
@slava77 , problem is that cmsdist CMSSW_10_2_X branch is very old and do not support testing external data PRs. If cms-data/RecoBTag-Combined#29 looks good then we can merge it first and provide a cmsdist PR directly to test with this cmssw pr.

**Comment by peruzzim** [comment]
@slava77 for the moment this should only be added under the run2_miniAOD_devel. For NanoAOD there will be a separate discussion later, but the introduction of a dedicated modifier is not foreseen.

File: PhysicsTools/PatAlgos/python/tools/jetTools.py [hunk 2]
```diff
@@ -678,13 +664,36 @@ def setupBTagging(process, jetSource, pfCandidates, explicitJTA, pvSource, svSou
...
+                else:
+                    raise ValueError("Invalid pfCandidates collection: %s." % pfCandidates.value())
+                addToProcessAndTask(btagPrefix+btagInfo+labelName+postfix,
+                                    btag.pfParticleNetTagInfos.clone(
+                                      jets = jetSource,
+                                      vertices = pvSource,
+                                      secondary_vertices = svSource,
+                                      pf_candidates = pfCandidates,
+                                      puppi_value_map = puppi_value_map,
+                                      vertex_associator = vertex_associator,
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/tools/jetTools.py:693]: while trying to run this in re-miniAOD (AOD input) the job fails apparently related to ```fix_daughters```.
This part should be updated here.

**Comment by smuzaffar** [comment]
test parameter
 - pull_request = cms-sw/cmsdist#5719

**Comment by smuzaffar** [comment]
please test
@slava77 cmsdist Pr to go with this is https://github.com/cms-sw/cmsdist/pull/5719

**Comment by santocch** [comment]
+1

**Comment by slava77** [comment]
@hqucms 
please clarify on the status/plan for updating this PR:
- use ```run2_miniAOD_devel```
- fix running on AOD inputs, likely by changing ```fix_daughters```

Thank you.

**Comment by hqucms** [comment]
@slava77 Done in https://github.com/cms-sw/cmssw/pull/29063/commits/fcc6974c2137f34f0d8b68d6af080555afa9c1fb and https://github.com/cms-sw/cmssw/pull/29063/commits/0beb8abfe224d11aa3ab1567acc05c7f6caecf36.

**Comment by slava77** [comment]
@cmsbuild please test

this will just check that the "production" workflows are not broken; I'll re-check locally for the test with _devel enabled

**Comment by santocch** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/29063	0beb8ab 
- code changes are in line with the PR description and the follow up review. The backport is consistent with the master version, ignoring some parts that do not apply for the backport (e.g. BoostedJetMXNetJetTagsProducer.cc fillDescriptions and the coupled pfParticleNet_cff.py)
- jenkins tests pass and comparisons with the baseline show no differences (by default the new code is not running)
- the PR was tested locally using 1325.7 and 136.8311 with run2_miniAOD_devel modifier added to enable the non-production extras: changes appear only in slimmedJetsAK8 in the jet tag discriminants with only particleNet taggers  added  while the available ones are unmodified; a "remote" comparison with the same workflow 136.8311 in 11_1_0_pre5 shows fairly similar results in the particleNet tagger values.

**Comment by silviodonato** [comment]
+1

**Pull request merged**